crates/uv/tests/lock_scenarios.rs

```diff
@@ -0,0 +1,803 @@
+/// resulted in multiple versions of `a` being unconditionally included in the lock
+/// file. So this acts as a regression test to ensure that only one version of `a`
+/// is selected.
+///
+/// ```text
+/// fork-marker-selection
+/// ├── environment
+/// │   └── python3.8
+/// ├── root
+/// │   ├── requires a
+/// │   │   ├── satisfied by a-0.1.0
+/// │   │   └── satisfied by a-0.2.0
+/// │   ├── requires b>=2; sys_platform == "linux"
+/// │   │   └── satisfied by b-2.0.0
+/// │   └── requires b<2; sys_platform == "darwin"
+/// │       └── satisfied by b-1.0.0
+/// ├── a
+/// │   ├── a-0.1.0
+/// │   └── a-0.2.0
+/// │       └── requires b>=2.0.0
+/// │           └── satisfied by b-2.0.0
+/// └── b
+///     ├── b-1.0.0
+///     └── b-2.0.0
+/// ```
+#[test]
+fn fork_marker_selection() -> Result<()> {
+    let context = TestContext::new("3.8");
+
+    // In addition to the standard filters, swap out package names for shorter messages
+    let mut filters = context.filters();
+    filters.push((r"fork-marker-selection-", "package-"));
+
+    let pyproject_toml = context.temp_dir.child("pyproject.toml");
+    pyproject_toml.write_str(
+        r###"
+        [project]
+        name = "project"
+        version = "0.1.0"
+        dependencies = [
+          '''fork-marker-selection-a''',
+          '''fork-marker-selection-b>=2; sys_platform == "linux"''',
+          '''fork-marker-selection-b<2; sys_platform == "darwin"''',
+        ]
+        "###,
+    )?;
+
+    let mut cmd = context.lock_without_exclude_newer();
+    cmd.arg("--index-url")
+        .arg("https://astral-sh.github.io/packse/0.3.17/simple-html/");
+    uv_snapshot!(filters, cmd, @r###"
+    success: true
+    exit_code: 0
+    ----- stdout -----
+
+    ----- stderr -----
+    warning: `uv lock` is experimental and may change without warning.
+    Resolved 5 packages in [TIME]
+    "###
+    );
+
+    let lock = fs_err::read_to_string(context.temp_dir.join("uv.lock"))?;
+    insta::with_settings!({
+        filters => context.filters(),
+    }, {
+        assert_snapshot!(
+            lock, @r###"
+        version = 1
+
+        [[distribution]]
+        name = "fork-marker-selection-a"
+        version = "0.1.0"
+        source = "registry+https://astral-sh.github.io/packse/0.3.17/simple-html/"
+        sdist = { url = "https://astral-sh.github.io/packse/0.3.17/files/fork_marker_selection_a-0.1.0.tar.gz#sha256=03c464276ee75f5a1468da2a4090ee6b5fda0f26f548707c9ffcf06d3cf69282", hash = "sha256:03c464276ee75f5a1468da2a4090ee6b5fda0f26f548707c9ffcf06d3cf69282" }
+        wheels = [{ url = "https://astral-sh.github.io/packse/0.3.17/files/fork_marker_selection_a-0.1.0-py3-none-any.whl#sha256=0e45ca7b3616810a583dc9754b52b91c69aeea4070d6fe0806c67081d0e95473", hash = "sha256:0e45ca7b3616810a583dc9754b52b91c69aeea4070d6fe0806c67081d0e95473" }]
+
+        [[distribution]]
+        name = "fork-marker-selection-a"
+        version = "0.2.0"
```

Comment:
Note that this test result is wrong. Notice that `fork-marker-selection-a` is included twice here, with two different versions, but _no_ marker expressions. I believe this needs to be fixed by tracking marker expressions in forked resolver states. Since that isn't trivial to fix, and I'm not sure if there's a good way to mark packse tests as "wrong," I figured it would be best to just get this merged and then fix it in a follow-up.